dd-sdk-android-gradle-plugin/src/test/kotlin/com/datadog/gradle/plugin/internal/VariantIteratorTest.kt

```diff
@@ -0,0 +1,140 @@
+import org.mockito.junit.jupiter.MockitoExtension
+import org.mockito.junit.jupiter.MockitoSettings
+import org.mockito.quality.Strictness
+
+@Extensions(
+    ExtendWith(MockitoExtension::class),
+    ExtendWith(ForgeExtension::class)
+)
+@MockitoSettings(strictness = Strictness.LENIENT)
+@ForgeConfiguration(Configurator::class)
+internal class VariantIteratorTest {
+
+    @Test
+    fun `𝕄 iterate 𝕎 forEach() { 1 dimension }`(
+        @StringForgery(case = Case.LOWER) a: String
+    ) {
+        // Given
+        val iterator = VariantIterator(listOf(a))
+
+        // When
+        val output = mutableListOf<String>()
+        iterator.forEach { output.add(it) }
+
+        // Then
+        assertThat(output)
+            .containsExactly(a)
+    }
+
+    @Test
+    fun `𝕄 iterate 𝕎 forEach() { 2 dimensions }`(
+        @StringForgery(case = Case.LOWER) a: String,
+        @StringForgery(case = Case.LOWER) b: String
+    ) {
+        // Given
+        val B = b.capitalize()
+        val iterator = VariantIterator(listOf(a, b))
+
+        // When
+        val output = mutableListOf<String>()
+        iterator.forEach { output.add(it) }
+
+        // Then
+        assertThat(output)
+            .containsExactly(b, a, a + B)
+    }
+
+    @Test
+    fun `𝕄 iterate 𝕎 forEach() { 3 dimensions }`(
+        @StringForgery(case = Case.LOWER) a: String,
+        @StringForgery(case = Case.LOWER) b: String,
+        @StringForgery(case = Case.LOWER) c: String
+    ) {
+        // Given
+        val B = b.capitalize()
+        val C = c.capitalize()
+        val iterator = VariantIterator(listOf(a, b, c))
+
+        // When
+        val output = mutableListOf<String>()
+        iterator.forEach { output.add(it) }
+
+        // Then
+        assertThat(output)
+            .containsExactly(c, b, b + C, a, a + C, a + B, a + B + C)
+    }
+
+    @Test
+    fun `𝕄 iterate 𝕎 forEach() { 4 dimensions }`(
+        @StringForgery(case = Case.LOWER) a: String,
+        @StringForgery(case = Case.LOWER) b: String,
+        @StringForgery(case = Case.LOWER) c: String,
+        @StringForgery(case = Case.LOWER) d: String
+    ) {
+        // Given
+        val B = b.capitalize()
+        val C = c.capitalize()
+        val D = d.capitalize()
+        val iterator = VariantIterator(listOf(a, b, c, d))
+
+        // When
+        val output = mutableListOf<String>()
+        iterator.forEach { output.add(it) }
+
+        // Then
+        assertThat(output)
+            .containsExactly(
+                d, c, c + D,
+                b, b + D, b + C, b + C + D,
+                a, a + D, a + C, a + C + D,
+                a + B, a + B + D, a + B + C,
+                a + B + C + D
+            )
+    }
+
+    @Test
+    fun `𝕄 iterate 𝕎 forEach() { anySize }`(
+        @StringForgery(case = Case.LOWER) names: List<String>
+    ) {
+        // Given
+        val limitedNames = names.take(10) // limit the complexity of the data
```

Comment:
that is a good point. `1 << names.size` can grow quickly, especially for big apps and can slow down (or even break) the build. wondering how to find the information how many dimensions can exist in the big apps

Reply:
I have a guess that even for big apps, the number of flavor dimensions should stay reasonable as it will also impact gradle's build time (not to mention the complexity for engineers to keep track of all those dimensions in their mind when writing their app)